### PR TITLE
[JENKINS-34534] support for gitlab-plugin 1.2.0

### PR DESCRIPTION
**Pull request by pyther**
- `allowAllBraches` paramater was replaced by `branchFilterType`
- new parameter `targetBranchRegex`
- new property `gitlabConnection` which sets the gitlab connection a job
  should use

### CODE REVIEW

**Comment by daspilker** [comment]
This breaks compatibility for users who are still using version 1.1.x of the GitLab plugin. Keep the existing options for compatibility and only generate new option when a newer version of the plugin is installed. See https://github.com/jenkinsci/job-dsl-plugin/blob/job-dsl-1.46/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy#L203 for an example.

**Comment by MrTrustor** [comment]
For anyone having problems with DSL and the Gitlab Plugin 1.2.x, reverting to Gitlab plugin 1.1.32 is a temporary fix.

**Comment by pyther** [comment]
@daspilker I have update the changeset per your request. I stumbled around for a bit, but I think I got on the right track. I would appreciate if you could review the change again.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/properties/PropertiesContext.groovy [hunk 1]
```diff
@@ -173,4 +173,18 @@ class PropertiesContext extends AbstractExtensibleContext {
...
+
+    /**
+     * Configures the Gitlab Connection to use
+     *
+     * Gitlab connections can be configured under global settings
+     *
+     * @since 1.46
+     */
+    @RequiresPlugin(id = 'gitlab-plugin', minimumVersion = '1.2.0')
+    void gitlabConnection(String connectionName) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/properties/PropertiesContext.groovy]: remove this amd use the [automatically generated DSL](https://github.com/jenkinsci/job-dsl-plugin/wiki/Automatically-Generated-DSL) for new stuff

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/TriggerContext.groovy [hunk 2]
```diff
@@ -202,8 +202,10 @@ class TriggerContext extends ItemTriggerContext {
      *
      * @since 1.42
      */
-    @RequiresPlugin(id = 'gitlab-plugin', minimumVersion = '1.1.28')
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/TriggerContext.groovy]: do not remove the minimum version requirements, it's still valid

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/triggers/TriggerContextSpec.groovy [hunk 3]
```diff
@@ -353,14 +353,17 @@ class TriggerContextSpec extends Specification {
     }
 
     def 'call gitlabPush trigger with no options'() {
+        setup:
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/triggers/TriggerContextSpec.groovy]: leave this test as it is and add a new test for 1.2.0

**Comment by daspilker** [comment]
OK, cool. See my inline comments. And mark `allowAllBranches` as deprecated.

**Comment by pyther** [comment]
@daspilker updated. Let me know if there is anything else that needs to be updated/changed.

**Comment by pyther** [comment]
@daspilker I'm a bit confused on behavior I'm seeing. I can swear this worked yesterday.

I updated a jenkins and bunch of plugins. Additionally, I built the job-dsl-plugin from master.

DSL snippet
```
job.with {
  triggers {
    gitlabPush {
      buildOnMergeRequestEvents(false)
      buildOnPushEvents(true)
      enableCiSkip(true)
      setBuildDescription(true)
      addNoteOnMergeRequest(false)
      rebuildOpenMergeRequest('never')
      addVoteOnMergeRequest(false)
      useCiFeatures(true)
      branchFilterType('nameBasedFilter')
}
```

When I run my job-dsl job, I get the following error:
```
ERROR: (Trigger.groovy, line 19) No signature of method: javaposse.jobdsl.dsl.helpers.triggers.GitLabTriggerContext.branchFilterType() is applicable for argument types: (java.lang.String) values: [nameBasedFilter]
Possible solutions: setBranchFilterType(java.lang.String), getBranchFilterType()
```

If I change `branchFilterType('nameBasedFilter')` to `setBranchFilterType('nameBasedFilter')` per the recommendation, everything works.

The API Viewer shows `branchFilterType`. As a sanity check I grepped the code for `setBranchFilterType` and didn't get any matches.

Do you have any ideas on what might be going on?

**Comment by daspilker** [comment]
I did some refactoring while merging, `branchFilterType` will be set implicitly, see

https://github.com/jenkinsci/job-dsl-plugin/blob/master/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/GitLabTriggerContext.groovy#L36
https://github.com/jenkinsci/job-dsl-plugin/blob/master/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/GitLabTriggerContext.groovy#L44
https://github.com/jenkinsci/job-dsl-plugin/blob/master/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/GitLabTriggerContext.groovy#L57

**Comment by pyther** [comment]
I see, makes sense. I was afraid I submitted busted code!

I didn't notice before, but there are two triggers in the api viewer:`gitlabPush` and `gitLabPushTrigger`.

`gitLabPushTrigger` is automatically generated. Some of the parameter names differ. Should there be any efforts to fix/update the names in the job-dsl code or will the auto generated functions be preferred?

**Comment by daspilker** [comment]
For compatibility I'm not going to adjust the built-in DSL names.

The plan is to focus on improving the plugins (e.g. the GitLab plugin) so that the automatically generated DSL improves. And other DSLs which are use the Structs plugin, e.g. the Pipeline DSL.

One thing that needs to be improved is that most plugins use `@DataBoundConstructor` to declare all options. But `@DataBoundConstructor` defines mandatory options and in post cases the options are not mandatory and should use `@DataBoundSetter` instead. So currently you need to specify all options when using `gitLabPushTrigger`.

**Pull request merged**